### DISCUSSION TITLE
Use with SvelteKit - need to use [ssrExchange]?

### DISCUSSION THREAD

**Discussion by nimmolo** [urql-graphql/urql#2053] [Q&A] [answered]
Hi, i'm trying to get my first urql query working, posting from a SvelteKit app page at `localhost:3001/observations` to a Rails back end at `localhost:3000/graphql`. I realize both SvelteKit and the `urql-svelte` code are new and in flux, but it seems some others have got this basic kind of query working(?)

(Note: previously I attempted this with apollo-client. The request did hit the `localhost:3000/graphql` endpoint and the Rails terminal showed it executed the desired query, implying that my Rails routes are set up OK. I couldn't get it to return any data, though, and after a few days not getting it to work I'm trying out urql.)

When I try the urql query, the Rails server is throwing an error it did not throw with the apollo-client request: `ActionController::RoutingError (No route matches [OPTIONS] "/graphql")`. This error can imply that the request was not a POST method -- so I explicitly set `preferGetMethod: false`. I doublechecked my Rails routes and it is set to route `post` actions at endpoint `/graphql` to `/graphql#execute`, this is correct.

Here's my code in `routes/observations/index.svelte`:

```
<script lang="ts">
  import { createClient, setClient, operationStore, query } from '@urql/svelte';

  const client = createClient({
    url: 'http://localhost:3000/graphql',
	preferGetMethod: false
    });

    const observations = operationStore(`
      query {
        observations {
          pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
          }
          edges {
            cursor
            node {
              id
              textName
            }
          }
        }
      }
  `);

  setClient(client);

  query(observations);
</script>

{#if $observations.fetching}
  Loading...
{:else if $observations.error}
  Error: {$observations.error.message}
{:else}
  <script>
    console.log($observations);
  </script>
{/if}
```

In SvelteKit it's returning the error `Error: [Network] Preflight response is not successful` (while Rails reports `No route matches [OPTIONS] "/graphql")` as noted above). 

What's going on here? Do i need to be using `dedupExchange, cacheExchange, fetchExchange, ssrExchange`  as explained in the urql API docs?

**Comment by armchair-traveller** [comment, marked as answer]
You need to respond to CORS preflight on your rails backend. This isn't really a GraphQL, Svelte Kit, or urql issue, it's just preflight which is part of web standards. See https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request

Essentially you need to return a preflight response (basically says the server accepts modern request options) when an OPTIONS request is issued 
> A preflight request is automatically issued by a browser

Because of this, it means the response shouldn't contain data besides a simple acknowledgement as the preflight request doesn't have your GraphQL query.
You should also handle CORS after the preflight. (details also linked in the MDN preflight page, under **See also**)

You don't have to deal with any of this if it's on the same site. Your backend is decoupled from Svelte Kit. If it's using a Svelte Kit endpoint you'd normally be on the same site anyway (and it'd be how you'd handle SSR). But you can also proxy your rails server. In that case, you'd assign the same domain name for both the backend and frontend for prod.

**Reply by nimmolo** [reply]
Should have known this! ...sorry and thanks for responding!
